src/ghci/mod.rs

```diff
@@ -0,0 +1,351 @@
+//! The core [`Ghci`] session struct.
+
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::process::Stdio;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+
+use camino::Utf8PathBuf;
+use itertools::Itertools;
+use miette::IntoDiagnostic;
+use miette::WrapErr;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::BufReader;
+use tokio::process::Child;
+use tokio::process::ChildStderr;
+use tokio::process::Command;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use tokio::sync::Mutex;
+use tokio::task;
+use tokio::task::JoinHandle;
+use tracing::instrument;
+
+mod stdin;
+use stdin::GhciStdin;
+use stdin::StdinEvent;
+
+mod stdout;
+use stdout::GhciStdout;
+use stdout::StdoutEvent;
+
+mod show_modules;
+use show_modules::ModuleSet;
+
+use crate::buffers::LINE_BUFFER_CAPACITY;
+use crate::event_filter::FileEvent;
+use crate::incremental_reader::IncrementalReader;
+use crate::sync_sentinel::SyncSentinel;
+
+/// The `ghci` prompt we use. Should be unique enough, but maybe we can make it better with Unicode
+/// private-use-area codepoints or something in the future.
+pub const PROMPT: &str = "###~GHCID-NG-PROMPT~###";
+
+/// The name we import `System.IO` as in `ghci`. This is used to run a few `putStrLn` commands and
+/// similar without messing with the user's namespace. If you have a module in your project named
+/// `GHCID_NG_IO_INTERNAL__` that's on you.
+pub const IO_MODULE_NAME: &str = "GHCID_NG_IO_INTERNAL__";
+
+/// A `ghci` session.
+pub struct Ghci {
+    /// A function which returns the command used to start this `ghci` session.
+    /// This needs to be an [`Arc`] because [`Command`] doesn't implement [`Clone`] and we need to
```

Comment:
Here's essentially what we want to write:

```rust
use std::process::Command;

struct Ghci {
    command: Command
}

impl Ghci {
    fn new(command: Command) -> Self {
        Self { command }
    }
}

fn main() {
    let command = Command::new("ghci");
    let mut ghci = Ghci::new(command);
    let new_ghci = Ghci::new(ghci.command);
    std::mem::replace(&mut ghci, new_ghci);
}
```

```
error[E0382]: borrow of partially moved value: `ghci`
  --> src/main.rs:17:23
   |
16 |     let new_ghci = Ghci::new(ghci.command);
   |                              ------------ value partially moved here
17 |     std::mem::replace(&mut ghci, new_ghci);
   |                       ^^^^^^^^^ value borrowed here after partial move
   |
   = note: partial move occurs because `ghci.command` has type `Command`, which does not implement the `Copy` trait
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e4dbe9ac608375b926bb984c4128e87b

Another way to solve this would be by keeping the `Ghci` value in an `Option<Ghci>`, so we could `.take()` the wrapped `Ghci` out of it without putting the `Option` in an invalid state:

```rust
let command = Command::new("ghci");
let mut ghci = Some(Ghci::new(command));
// Here, we take the value out of `ghci`, leaving `ghci` set to `None`:
let new_ghci = Some(Ghci::new(ghci.take().unwrap().command));
// Now, we can take an `&mut` reference to `ghci` because it doesn't contain the `Command` we took out:
std::mem::replace(&mut ghci, new_ghci);
```